### DISCUSSION TITLE
Common Troubleshooting

### DISCUSSION THREAD

**Discussion by 2percentsilk** [community/community#3434] [Codespaces]
## Cannot create a PR for upstream fork / cannot push to other repo
Currently, you can only push changes to the repository the codespace was created from or your personal repos (see our [docs](https://docs.github.com/en/github/developing-online-with-codespaces/managing-access-and-security-for-codespaces) for how to enable access to other personal repos).

We're working on refining this workflow but in the meantime, you can follow these steps as a workaround:
1. Create a [Personal Access Token](https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token#creating-a-token) with the following permissions:
  - `public_repo`
  - `read:discussion`
  - `read:org`
  - Depending on the permissions of the upstream repo, you may need additional permissions but the resulting terminal error should let you know which ones! 
2. Set the environment variable `GITHUB_TOKEN` to equal your PAT.
  - You can do this by running `export GITHUB_TOKEN='PAT'` in your terminal.

You should now be able to create PRs against an upstream fork and push to other repositories from the terminal.

References: #3408

**Comment by bartr** [comment]
i added this command to my dotfiles

` export GITHUB_TOKEN=$PAT` and it works great (when dotfiles are working)

**Comment by Chuxel** [comment]
Note that we do not recommend upgrading permissions in dotfiles due to the added risk this can creates. Every codespace for the repo you have granted access to the secret can now get access to this token using the common `GITHUB_TOKEN` env var. You loose the option to review for the repo before granting broader access to this commonly used variable - its probably okay when you are creating PRs, but the dotfiles aren't limited to PR scenarios, so it becomes more of a repository setting than a codespace setting.

Now, the problem is just doing this in a terminal as described above is it goes away when you close the terminal. Instead, if you need it to persist, you can run the following script that swaps out the credential manager for one that uses your own token only in the codespace you actually need it. **We don't recommend doing this broadly**, but it lets you opt-into doing it on a codespace-by-codespace basis so it only kicks when you actually need it. 

```bash
sudo sed -i -E 's/helper =.*//' /etc/gitconfig
git config --global credential.helper '!f() { sleep 1; echo "username=${GH_USER}"; echo "password=${GH_TOKEN}"; }; f'
```

You can then set user secrets with `GH_USER` and `GH_TOKEN` (or whatever you select).

**Reply by bartr** [reply]
what does "every repo you clone can get access to this token" mean?  I thought my PAT just controlled "logging into" the repo?

**Reply by bartr** [reply]
$PAT is a Codespace secret, so unless I explicitly trust the repo and tell it to send the secrets, the env var won't exist in that Codespace - correct?  My install has an if -z $PAT, so it only gets changed if the PAT secret is set on Codespace creation.

**Reply by Chuxel** [reply]
@bartr Yeah clarified a bit with an edit... the added risk is that the `GITHUB_TOKEN` is the name any malicious code would be looking for and that doing that in a repo level does not give you the opportunity to investigate first and then just "opt-in" when you actually need it for a specific codespace.

**Reply by bartr** [reply]
duh - makes perfect sense.

I tested the above today and it works well. Now I need to adjust to use $PAT. I changed it slightly to use
`sudo git config --system ...` and it updated the credential.helper as expected.
